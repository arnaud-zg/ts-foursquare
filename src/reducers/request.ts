import { createReducer, getType } from 'typesafe-actions'
import {
  NEntity,
  NLikes,
  NMiniVenue,
  NRecommendedPlaces,
  NRequestState,
  NVenue,
  NVenueListed,
  NVenuesCategories,
} from '../../types'
import { TRootAction } from '../actions'
import {
  getVenuesCategoriesAsync,
  getVenuesExploreAsync,
  getVenuesLikesAsync,
  getVenuesListedAsync,
  getVenuesNextVenuesAsync,
  getVenuesSearchAsync,
  getVenuesSimilarAsync,
  getVenuesSuggestCompletionAsync,
  getVenuesTrendingAsync,
} from '../actions/venues'
import { ASYNC_ACTION_NAME_MAPPING } from '../constants/asyncAction'
import { getVenuesListedGroupKey } from '../utils/venue'

export const initialState: NRequestState.IState = {}

export const requestReducer = createReducer<NRequestState.IState, TRootAction>(
  initialState
)
  .handleAction(
    getType(getVenuesCategoriesAsync.success),
    (state, action): NRequestState.IState => {
      const entityIds = action.payload.map(
        (item: NVenuesCategories.ICategory) => item.id
      )

      return {
        ...state,
        [ASYNC_ACTION_NAME_MAPPING[action.type]]: {
          entityIds,
        },
      }
    }
  )
  .handleAction(
    getType(getVenuesExploreAsync.success),
    (state, action): NRequestState.IState => {
      const entityIds = action.payload.map(
        (item: NRecommendedPlaces.IGroupItem) => item.venue.id
      )

      return {
        ...state,
        [ASYNC_ACTION_NAME_MAPPING[action.type]]: {
          entityIds,
        },
      }
    }
  )
  .handleAction(
    getType(getVenuesLikesAsync.success),
    (state, action): NRequestState.IState => {
      const entityIds = action.payload.items.map(
        (item: NLikes.IItem) => item.id
      )

      return {
        ...state,
        [ASYNC_ACTION_NAME_MAPPING[action.type]]: {
          entityIds,
        },
      }
    }
  )

  .handleAction(
    getType(getVenuesListedAsync.success),
    (state, action): NRequestState.IState => {
      const entityIds = action.payload.groups.map(
        (group: NEntity.IEntityGroup<NVenueListed.IGroupItem>) =>
          getVenuesListedGroupKey(group)
      )

      return {
        ...state,
        [ASYNC_ACTION_NAME_MAPPING[action.type]]: {
          entityIds,
        },
      }
    }
  )

  .handleAction(
    [
      getType(getVenuesNextVenuesAsync.success),
      getType(getVenuesSearchAsync.success),
      getType(getVenuesSimilarAsync.success),
      getType(getVenuesTrendingAsync.success),
    ],
    (state, action): NRequestState.IState => {
      const entityIds = action.payload.map((item: NVenue.IVenue) => item.id)

      return {
        ...state,
        [ASYNC_ACTION_NAME_MAPPING[action.type]]: {
          entityIds,
        },
      }
    }
  )
  .handleAction(
    getType(getVenuesSuggestCompletionAsync.success),
    (state, action): NRequestState.IState => {
      const entityIds = action.payload.map(
        (item: NMiniVenue.IMiniVenue) => item.id
      )

      return {
        ...state,
        [ASYNC_ACTION_NAME_MAPPING[action.type]]: {
          entityIds,
        },
      }
    }
  )
  .handleAction(
    [
      getType(getVenuesCategoriesAsync.cancel),
      getType(getVenuesExploreAsync.cancel),
      getType(getVenuesLikesAsync.cancel),
      getType(getVenuesNextVenuesAsync.cancel),
      getType(getVenuesSearchAsync.cancel),
      getType(getVenuesSimilarAsync.cancel),
      getType(getVenuesSuggestCompletionAsync.cancel),
      getType(getVenuesTrendingAsync.cancel),
      getType(getVenuesListedAsync.cancel),
    ],
    (state, action): NRequestState.IState => ({
      ...state,
      [ASYNC_ACTION_NAME_MAPPING[action.type]]: {
        cancelReason: action.payload,
        entityIds: [],
      },
    })
  )
  .handleAction(
    [
      getType(getVenuesCategoriesAsync.failure),
      getType(getVenuesExploreAsync.failure),
      getType(getVenuesLikesAsync.failure),
      getType(getVenuesNextVenuesAsync.failure),
      getType(getVenuesSearchAsync.failure),
      getType(getVenuesSimilarAsync.failure),
      getType(getVenuesSuggestCompletionAsync.failure),
      getType(getVenuesTrendingAsync.failure),
      getType(getVenuesListedAsync.failure),
    ],
    (state, action): NRequestState.IState => ({
      ...state,
      [ASYNC_ACTION_NAME_MAPPING[action.type]]: {
        entityIds: [],
        error: action.payload,
      },
    })
  )
