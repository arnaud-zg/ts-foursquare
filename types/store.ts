import { NLife } from './lifeState'
import { NStatus } from './statusState'
import { NVenues } from './venuesState'

export declare namespace NStore {
  interface IState {
    life: NLife.IState
    status: NStatus.IState
    venues: NVenues.IState
  }
}
