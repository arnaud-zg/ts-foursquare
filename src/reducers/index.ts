import { NVenues } from '../../types/venues'

export type TRootState = {
  venues: NVenues.IVenuesState
}
