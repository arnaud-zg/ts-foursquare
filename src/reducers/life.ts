import { createReducer } from 'typesafe-actions'
import { NLifeState } from '../../types'
import { putCredentials, TLifeAction } from '../actions'

export const initialState: NLifeState.IState = {
  status: false,
  credentials: {
    clientId: '',
    clientSecret: '',
  },
}

export const lifeReducer = createReducer<NLifeState.IState, TLifeAction>(
  initialState
).handleAction(
  putCredentials,
  (state, action): NLifeState.IState => ({
    ...state,
    status: true,
    credentials: {
      ...action.payload,
    },
  })
)
