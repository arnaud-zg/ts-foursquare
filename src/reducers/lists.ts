import { createReducer } from 'typesafe-actions'
import { NListsState } from '../../types'
import { getListsAsync, TListsAction } from '../actions'

export const initialState: NListsState.IState = {
  entities: {},
}

export const listsReducer = createReducer<NListsState.IState, TListsAction>(
  initialState
).handleAction(
  getListsAsync.success,
  (state, action): NListsState.IState => {
    return {
      ...state,
      entities: {
        ...state.entities,
        [action.payload.id]: action.payload,
      },
    }
  }
)
