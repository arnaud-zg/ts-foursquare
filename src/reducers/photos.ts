import { createReducer, getType } from 'typesafe-actions'
import { NPhotosState } from '../../types'
import { getPhotosDetailsAsync, TPhotosAction } from '../actions'

export const initialState: NPhotosState.IState = {
  entities: {},
}

export const photosReducer = createReducer<NPhotosState.IState, TPhotosAction>(
  initialState
).handleAction(
  getType(getPhotosDetailsAsync.success),
  (state, action): NPhotosState.IState => {
    return {
      ...state,
      entities: {
        ...state.entities,
        [action.payload.id]: action.payload,
      },
    }
  }
)
