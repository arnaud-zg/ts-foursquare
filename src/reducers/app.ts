import { combineReducers } from 'redux'
import { initialState as initialStateLife } from '../../src/reducers/life'
import { initialState as initialStateStatus } from '../../src/reducers/status'
import { initialState as initialStateVenues } from '../../src/reducers/venues'
import { lifeReducer } from './life'
import { statusReducer } from './status'
import { venuesReducer } from './venues'

export const initialState = {
  life: initialStateLife,
  status: initialStateStatus,
  venues: initialStateVenues,
}

export type TState = typeof initialState

export const appReducer = combineReducers({
  life: lifeReducer,
  status: statusReducer,
  venues: venuesReducer,
})
