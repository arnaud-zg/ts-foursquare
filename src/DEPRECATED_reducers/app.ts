import { combineReducers } from 'redux'
import { initialState as initialStateLife, lifeReducer } from './life'
import { initialState as initialStateLists, listsReducer } from './lists'
import { initialState as initialStatePhotos, photosReducer } from './photos'
import { initialState as initialStateRequest, requestReducer } from './request'
import { initialState as initialStateStatus, statusReducer } from './status'
import { initialState as initialStateVenues, venuesReducer } from './venues'

export const initialState = {
  life: initialStateLife,
  lists: initialStateLists,
  photos: initialStatePhotos,
  request: initialStateRequest,
  status: initialStateStatus,
  venues: initialStateVenues,
}

export type TState = typeof initialState

export const appReducer = combineReducers({
  life: lifeReducer,
  lists: listsReducer,
  photos: photosReducer,
  request: requestReducer,
  status: statusReducer,
  venues: venuesReducer,
})
