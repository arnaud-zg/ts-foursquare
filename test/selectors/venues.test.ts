import { initialState } from '../../src/reducers/app'
import {
  venueCategoriesStateSelector,
  venueLocationSelector,
  venuesCategoriesSelector,
  venueSelector,
  venuesSelector,
} from '../../src/selectors/venues'
import { NStore } from '../../types/store'
import { payload } from '../epics/__mocks__/getVenuesSearchAsync.resolve'

const venueId = payload.response.venues[0].id
const state: NStore.IState = {
  ...initialState,
  venues: {
    ...initialState.venues,
    entities: {
      [venueId]: payload.response.venues[0],
    },
  },
}

describe('selectors/venues', () => {
  it('should get venues', () => {
    expect(Object.keys(venuesSelector(state)).length).toEqual(1)
    expect(venuesSelector(state)).toMatchSnapshot()
  })

  it('should get a list a of categories', () => {
    expect(venuesCategoriesSelector(state)).toMatchSnapshot()
  })

  it('should get a specific venue', () => {
    expect(venueSelector(state, venueId)).toMatchSnapshot()
  })

  it('should get categories of a venue', () => {
    expect(venueCategoriesStateSelector(state, venueId)).toMatchSnapshot()
  })

  it('should get location of a venue', () => {
    expect(venueLocationSelector(state, venueId)).toMatchSnapshot()
  })
})
