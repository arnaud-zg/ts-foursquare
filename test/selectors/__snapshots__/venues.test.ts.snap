// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selectors/venues should get a list a of categories 1`] = `Array []`;

exports[`selectors/venues should get a specific venue 1`] = `
Object {
  "categories": Array [
    Object {
      "icon": Object {
        "prefix": "https://ss3.4sqi.net/img/categories_v2/food/steakhouse_",
        "suffix": ".png",
      },
      "id": "4bf58dd8d48988d1cc941735",
      "name": "Steakhouse",
      "pluralName": "Steakhouses",
      "primary": true,
      "shortName": "Steakhouse",
    },
  ],
  "hasPerk": false,
  "id": "3fd66200f964a5209beb1ee3",
  "location": Object {
    "address": "178 Broadway",
    "cc": "US",
    "city": "Brooklyn",
    "country": "United States",
    "crossStreet": "at Driggs Ave",
    "distance": 10,
    "formattedAddress": Array [
      "178 Broadway (at Driggs Ave)",
      "Brooklyn, NY 11211",
      "United States",
    ],
    "labeledLatLngs": Array [
      Object {
        "label": "display",
        "lat": 40.70995790682886,
        "lng": -73.96229110893742,
      },
    ],
    "lat": 40.70995790682886,
    "lng": -73.96229110893742,
    "postalCode": "11211",
    "state": "NY",
  },
  "name": "Peter Luger Steak House",
  "referralId": "v-1565543515",
  "venuePage": Object {
    "id": "77462637",
  },
}
`;

exports[`selectors/venues should get categories of a venue 1`] = `
Array [
  Object {
    "icon": Object {
      "prefix": "https://ss3.4sqi.net/img/categories_v2/food/steakhouse_",
      "suffix": ".png",
    },
    "id": "4bf58dd8d48988d1cc941735",
    "name": "Steakhouse",
    "pluralName": "Steakhouses",
    "primary": true,
    "shortName": "Steakhouse",
  },
]
`;

exports[`selectors/venues should get location of a venue 1`] = `
Object {
  "address": "178 Broadway",
  "cc": "US",
  "city": "Brooklyn",
  "country": "United States",
  "crossStreet": "at Driggs Ave",
  "distance": 10,
  "formattedAddress": Array [
    "178 Broadway (at Driggs Ave)",
    "Brooklyn, NY 11211",
    "United States",
  ],
  "labeledLatLngs": Array [
    Object {
      "label": "display",
      "lat": 40.70995790682886,
      "lng": -73.96229110893742,
    },
  ],
  "lat": 40.70995790682886,
  "lng": -73.96229110893742,
  "postalCode": "11211",
  "state": "NY",
}
`;

exports[`selectors/venues should get venues 1`] = `
Object {
  "3fd66200f964a5209beb1ee3": Object {
    "categories": Array [
      Object {
        "icon": Object {
          "prefix": "https://ss3.4sqi.net/img/categories_v2/food/steakhouse_",
          "suffix": ".png",
        },
        "id": "4bf58dd8d48988d1cc941735",
        "name": "Steakhouse",
        "pluralName": "Steakhouses",
        "primary": true,
        "shortName": "Steakhouse",
      },
    ],
    "hasPerk": false,
    "id": "3fd66200f964a5209beb1ee3",
    "location": Object {
      "address": "178 Broadway",
      "cc": "US",
      "city": "Brooklyn",
      "country": "United States",
      "crossStreet": "at Driggs Ave",
      "distance": 10,
      "formattedAddress": Array [
        "178 Broadway (at Driggs Ave)",
        "Brooklyn, NY 11211",
        "United States",
      ],
      "labeledLatLngs": Array [
        Object {
          "label": "display",
          "lat": 40.70995790682886,
          "lng": -73.96229110893742,
        },
      ],
      "lat": 40.70995790682886,
      "lng": -73.96229110893742,
      "postalCode": "11211",
      "state": "NY",
    },
    "name": "Peter Luger Steak House",
    "referralId": "v-1565543515",
    "venuePage": Object {
      "id": "77462637",
    },
  },
}
`;
